### ISSUE TITLE
Packagekit fails to remove packages to satisfy dependencies 

### ISSUE THREAD

**Issue by sidt4** [PackageKit/PackageKit#642]
Issue description:

`Packagekit` fails to remove packages to satisfy dependencies , where as `apt dist-upgrade` always succeeds in such cases.

apt upgrade works in 2 modes:

`[A]` - `apt-get upgrade` upgrades existing packages without installing / removing packages.
`[B]` - `apt-get dist-upgrade` upgrades existing packages + installs / removes packages.

`Packagekit` should either play safe with `[A]`, or should handle the challenges of `[B]`. Currently, it operates between `[A]` and `[B]` ( meaning it tries to install new packages as part of software update in GNOME Software, but doesn't remove the packages necessary for update to succeed ), which breaks software update in GNOME Software ( and possibly other clients too ) 

This issue was reported on Debian, but I think this applies to other package managers as well.

For more detailed information, please refer https://gitlab.gnome.org/GNOME/gnome-software/-/issues/2232

**Comment by pwithnall** [comment]
> This issue was reported on Debian, but I think this applies to other package managers as well.

I suspect the scope of this is actually limited to PackageKit’s apt backend. The other backends work differently due to the underlying package manager working differently.

**Comment by ximion** [comment]
This should actually work just fine with PackageKit and its APT backend - it certainly does with KDE Discover / Apper, last time I checked.
Does GNOME Software run Simulate* beforehand, and actually allow the backend to remove packages during a transaction?

**Comment by sidt4** [comment]
@ximion: If you take a look at https://gitlab.gnome.org/GNOME/gnome-software/-/issues/2232, there are 3 packages:

* libncurses5-dev
* libncursesw5-dev
* libtinfo-dev

 which don't want the upgrade to proceed as all of them have `libtinfo6 (= 6.4-4), libncurses-dev (= 6.4-4)` marked as their dependency. But, packages `libtinfo6-6.4+20230625-1` and `libncurses-dev-6.4+20230625-1` have to be installed as part of the upgrade.

Here, is the package information for `libtinfo6-6.4+20230625-1` and `libncurses-dev-6.4+20230625-1`.

```
$ apt-cache show libtinfo6

Package: libtinfo6
Source: ncurses
Version: 6.4+20230625-1
Installed-Size: 558
Maintainer: Ncurses Maintainers <ncurses@packages.debian.org>
Architecture: amd64
Depends: libc6 (>= 2.34)
Breaks: tmux (<< 3.3a-4)
Description-en: shared low-level terminfo library for terminal handling
 The ncurses library routines are a terminal-independent method of
 updating character screens with reasonable optimization.
 .
 This package contains the shared low-level terminfo library.
Description-md5: f681846d99e5156a0882bb53c35d3244
Multi-Arch: same
Homepage: https://invisible-island.net/ncurses/
Tag: role::shared-lib
Section: libs
Priority: optional
Filename: pool/main/n/ncurses/libtinfo6_6.4+20230625-1_amd64.deb
Size: 337796
MD5sum: 9b6d5713e7c0e24efab1049c8bd78a03
SHA256: 10a6c31a8e41743cc005e01aa9c82da8bd144159d9b9f3e1bb8dfd36090a936d
```

```
$ apt-cache show libncurses-dev

Package: libncurses-dev
Source: ncurses
Version: 6.4+20230625-1
Installed-Size: 2305
Maintainer: Ncurses Maintainers <ncurses@packages.debian.org>
Architecture: amd64
Replaces: ncurses-dev
Provides: libncurses5-dev (= 6.4+20230625-1), libncursesw5-dev (= 6.4+20230625-1), libtinfo-dev (= 6.4+20230625-1), ncurses-dev
Depends: libtinfo6 (= 6.4+20230625-1), libncurses6 (= 6.4+20230625-1), libncursesw6 (= 6.4+20230625-1), libc6-dev | libc-dev
Suggests: ncurses-doc
Conflicts: ncurses-dev
Description-en: developer's libraries for ncurses
 The ncurses library routines are a terminal-independent method of
 updating character screens with reasonable optimization.
 .
 This package contains the header files, static libraries
 and symbolic links that developers using ncurses will need.
Description-md5: 38da502b304ea3ed5ff1e12d072d14a8
Multi-Arch: same
Homepage: https://invisible-island.net/ncurses/
Tag: devel::library, role::devel-lib
Section: libdevel
Priority: optional
Filename: pool/main/n/ncurses/libncurses-dev_6.4+20230625-1_amd64.deb
Size: 347696
MD5sum: a8d4b374b1578e6b6ac390a2f5527ed6
SHA256: 7cffa4645d5a09dc28c7905ea8e84d1a8b17d66dd0f44b359e08572d32ad53f5
```

In this scenario, will `PackageKit` remove the 3 package mentioned above ?

**Comment by ximion** [comment]
It should. Can you try with `pkcon update --autoremove` to see if that works for you? If it does, then this is a GNOME Software bug, if it doesn't then the issue might be in PackageKit and we need to investigate this further here.

**Comment by sidt4** [comment]
`pkcon update` ( even without `--autoremove` ) works perfectly fine. I think `--autoremove`  is more of a "cleaning un-needed packages" after an update. So, it's not relevant here.

So, it appears to be an issue with gnome-software.

I'll update https://gitlab.gnome.org/GNOME/gnome-software/-/issues/2232 accordingly. This issue can be closed.

**Comment by ximion** [comment]
None of those options should have any impact...
My wild guess would have been that GS after running Simulate to retrieve the pending changes, somehow looses some state and instructs PK to perform a restricted action, instead of just performing the full upgrade. But I haven't looked at the code yet, so this might be completely wrong.

**Comment by mcrha** [comment]
As far as I can tell (see the call chain of the `gs_plugin_packagekit_update_apps_async`), the gnome-software gets list of all apps to be updated and if any is of the PackageKit "kind", it runs `pk_client_get_updates_async()` with no flags (`pk_bitfield_value (PK_FILTER_ENUM_NONE)`) and with `only-download=TRUE`. Once the `get-updates` finishes, there is called `pk_task_update_packages_async()` with the package IDs. When this succeeds, there is prepared an offline update.

Do you think that passing a list of the package IDs to the `pk_task_update_packages_async()` is the reason why the update from the gnome-software fails? The documentation suggests the package IDs are required.

With respect of the simulate process, it's used only as a `PK_TRANSACTION_FLAG_ENUM_SIMULATE` flag passed to the `pk_client_upgrade_system_async()`, which is used to get the list of failed package updates/upgrades.

**Comment by ximion** [comment]
> Do you think that passing a list of the package IDs to the `pk_task_update_packages_async()` is the reason why the update from the gnome-software fails? The documentation suggests the package IDs are required.

I think that's your problem! Is the set of package-ids passed in verbatim? If so, you disregard the simulated to-be state of the packages, e.g. some packages may not be in state "to be upgraded", but instead be in "to be removed".
The simplest possible thing would be something like this, if you don't allow the user to select individual package updates:

```c
static gboolean
update_system_filter_cb (PkPackage *package, gpointer user_data)
{
	PkInfoEnum package_enum = pk_package_get_info (package);
	return (package_enum != PK_INFO_ENUM_OBSOLETING && package_enum != PK_INFO_ENUM_REMOVING);
}

sack = pk_results_get_package_sack (results);
pk_package_sack_remove_by_filter (sack, &update_system_filter_cb, NULL);
``` 
(not tested, that's just how I remember it)

If GNOME Software actually does this, then I am still puzzled. But if this is missing, it would explain the issue.

**Comment by mcrha** [comment]
I do not see any such filtering in the code (briefly searched the code), but this is not the first time a package is replaced by another package and it worked so far - at least I'd expect this striking much sooner than in 2023. I even recall I saw the update history which contained removed packages, not only updated packages, thus that worked before for sure. There had been done changes in the related code on the gnome-software side, thus it's likely the changes caused a regression. I'll investigate it further and let you know. Thank you for the insights.

**Comment by sidt4** [comment]
> at least I'd expect this striking much sooner than in 2023.

@mcrha: Just curious. Is this a recent regression or was the issue always there.

**Comment by sidt4** [comment]
>  I still do not understand why it did not strike earlier

Ok. Did a search in gitlab gnome-software issues.

https://gitlab.gnome.org/GNOME/gnome-software/-/issues/951 seems to be related.

**Comment by mcrha** [comment]
> Is this a recent regression or was the issue always there.

Knowing that, I write it, I'm sorry. Maybe the right conditions were not met until know.

> Did a search in gitlab gnome-software issues.

That should happen there, in the gnome-software issues, I believe, not here.

**Comment by sidt4** [comment]
@ximion : Based on the comments in https://gitlab.gnome.org/GNOME/gnome-software/-/issues/2232#note_1790974, can you suggest the next steps ?

**Comment by ximion** [comment]
@sidt4 Given that PackageKit in its own tools calls this API differently and does not have this problem, and other clients like Apper, GNOME PackageKit and KDE Discover are also calling the API like pkcon does and are also not experiencing this issue, I think it's safe to say that GNOME Software is doing this wrong here and needs to be fixed (maybe PK could use some better documentation, of course).

Currently, GNOME Software also tells the backend "please upgrade `obsolete-package`" via its code, instead of `please remove `obsolete-package`" or simply "I don't request a specific action on `obsolete-package`". I think this is conceptually wrong, so the pkcon and other PK client's behavior makes a lot of sense. That DNF works is a lucky thing, but other backends may not be as lenient with upgrade requests, and it looks like the APT backend is one example of that, but there might be more.

tl;dr, I really think GNOME Software must be fixed here.

**Comment by mcrha** [comment]
> tl;dr, I really think GNOME Software must be fixed here.

There exists [a proposed patch](https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/1737) for the gnome-software, waiting for a review for a bit more than a week now. There is nothing to continue on here, from my point of view.
